[20903] Documentation to serialize DynamicData to JSON (#844)
eProsima/Fast-DDS-docs: docs/fastdds/xtypes/type_serializing.rst

.. include:: ../../03-exports/aliases-api.include

.. _xtypes_serialization_utilities:

Serialization Utilities
=======================

Fast DDS provides methods to serialize `XTypes` objects, enabling efficient data exchange and manipulation
within distributed systems.
Serialization is a crucial process in data distribution services, as it converts complex data structures into a
format that can be easily transmitted and reconstructed across different platforms and programming environments.

DynamicData to JSON
--------------------

In the context of DDS (Data Distribution Service), `DynamicType` represents the structure of the data being
distributed across the system.
Each `DynamicData` object corresponds to an object of the type represented by its `DynamicType`,
providing functionalities to access and modify data values.
To enhance interoperability and readability, it is necessary to serialize `DynamicData` into a more manageable format,
to enable easier data processing and analysis across different systems and applications.

> The method `json_serialize` converts a `DynamicData` object into an JSON object, then dumped into a `std::ostream`.

Reviewer: ``json_serialize`` should be a link to the API reference. 

```suggestion
The method ``json_serialize`` converts a `DynamicData` object into an JSON object, then dumped into a ``std::ostream``.
```

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_MAPS
    :end-before: //!--

The previous `DynamicData` object would be serialized as follows:

.. literalinclude:: /../code/json/Maps.json
    :language: json

Structure Types
"""""""""""""""

Structure types are used to group different types of data together.
Here is an example of the definition of structure types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_STRUCT
    :end-before: //!--

The previous `DynamicData` object would be serialized as follows:

.. literalinclude:: /../code/json/Structs.json
    :language: json

Union Types
"""""""""""

Union types are a special type of structure type where only one member exists.
Below is an example of the definition of union types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_UNION
    :end-before: //!--

The previous `DynamicData` object would be serialized as follows:

.. literalinclude:: /../code/json/Unions.json
    :language: json

> Example: Remote discovered types introspection

Reviewer: ```suggestion
Example: Convert received data into JSON format
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following code demonstrates how to implement remote type introspection using FastDDS in C++.
This feature allows a subscriber to introspect the remotely discovered type and serialize it into a more manageable
and understandable format.
Once a type is discovered, it can be registered and every time the subscriber receives new data related to this type,

> the corresponding DynamicData can be obtained from the DynamicDataFactory and serialized into a JSON string format.

Reviewer: The example should be focused on how to use the ``json_serialize`` function and not on how to implement type introspection. Also, add a link to the type introspection example just for completeness.

.. include:: ../../03-exports/aliases-api.include

.. _xtypes_serialization_utilities:

Serialization Utilities
=======================

Fast DDS provides methods to serialize `XTypes` objects, enabling efficient data exchange and manipulation
within distributed systems.
Serialization is a crucial process in data distribution services, as it converts complex data structures into a
format that can be easily transmitted and reconstructed across different platforms and programming environments.

DynamicData to JSON
--------------------

In the context of DDS (Data Distribution Service), |DynamicType-api| represents the structure of the data being
distributed across the system.
Each |DynamicData-api| object corresponds to an object of the type represented by its |DynamicType-api|,
providing functionalities to access and modify data values.
To enhance interoperability and readability, it is necessary to serialize |DynamicData-api| into a more manageable format,
to enable easier data processing and analysis across different systems and applications.
The method |XTypesUtils-json_serialize-api| converts a |DynamicData-api| object into an JSON object, then dumped into a ``std::ostream``.

> Supported Types Serialization

Reviewer: ```suggestion
Supported Types
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This section provides the serialization of |DynamicData-api| to JSON ostream for all supported types.

.. _xtypes_serialization_utilities_primitive:

Primitive types
"""""""""""""""

Primitive types are the basic building blocks for every |DynamicType-api|.
Below is an example of the definition of primitive types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_PRIMITIVES
    :end-before: //!--

The previous |DynamicData-api| object would be serialized as follows:

.. literalinclude:: /../code/json/Primitives.json
    :language: json

.. _xtypes_serialization_utilities_string:

> String types

Reviewer: ```suggestion
Strings
```

""""""""""""

String types are used to represent sequences of characters, which are essential for handling textual
data within the system.
The following example shows the definition of string types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_STRINGS
    :end-before: //!--

The previous |DynamicData-api| object would be serialized as follows:

.. literalinclude:: /../code/json/Strings.json
    :language: json

.. _xtypes_serialization_utilities_enumeration:

> Enumeration types

Reviewer: Apply the same to all types (plural and remove "types")

```suggestion
Enumerations
```

.. literalinclude:: /../code/json/Maps.json
    :language: json

.. _xtypes_serialization_utilities_structure:

Structure Types
"""""""""""""""

Structure types are used to group different types of data together.
Here is an example of the definition of structure types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_STRUCT
    :end-before: //!--

The previous |DynamicData-api| object would be serialized as follows:

.. literalinclude:: /../code/json/Structs.json
    :language: json

.. _xtypes_serialization_utilities_union:

Union Types
"""""""""""

Union types are a special type of structure type where only one member exists.
Below is an example of the definition of union types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_UNION
    :end-before: //!--

The previous |DynamicData-api| object would be serialized as follows:

.. literalinclude:: /../code/json/Unions.json
    :language: json

.. _xtypes_serialization_utilities_example:

Example: Convert received data into JSON format
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The following code demonstrates how to use the |XTypesUtils-json_serialize-api| function in FastDDS to serialize

Reviewer: ```suggestion
The following code demonstrates how to use the |XTypesUtils-json_serialize-api| function in Fast DDS to serialize
```

received data into a more manageable and understandable JSON format.
Each time the subscriber receives new data, the corresponding |DynamicData-api| can be obtained from the
|DynamicDataFactory-api| and serialized into a JSON string format.

> For more details on how to implement remote type discovery, see the :ref:`use-case-remote-type-discovery-and-matching`.

Reviewer: ```suggestion
Please refer to :ref:`use-case-remote-type-discovery-and-matching` section for more details on how to implement remote type discovery.
```

.. include:: ../../03-exports/aliases-api.include

.. _xtypes_serialization_utilities:

Serialization Utilities
=======================

Fast DDS provides methods to serialize `XTypes` objects, enabling efficient data exchange and manipulation
within distributed systems.
Serialization is a crucial process in data distribution services, as it converts complex data structures into a
format that can be easily transmitted and reconstructed across different platforms and programming environments.

DynamicData to JSON
--------------------

In the context of DDS (Data Distribution Service), |DynamicType-api| represents the structure of the data being
distributed across the system.
Each |DynamicData-api| object corresponds to an object of the type represented by its |DynamicType-api|,
providing functionalities to access and modify data values.

> To enhance interoperability and readability, it is necessary to serialize |DynamicData-api| into a more

Reviewer 2: ```suggestion
To enhance interoperability and readability, it is often useful to serialize |DynamicData-api| into a more
```

Reviewer 2: Or something similiar, "necessary" sounds a bit exaggerated to me.

manageable format, to enable easier data processing and analysis across different systems and applications.

> The method |XTypesUtils-json_serialize-api| converts a |DynamicData-api| object into an JSON object, then

Reviewer 2: ```suggestion
The method |XTypesUtils-json_serialize-api| converts a |DynamicData-api| object into a JSON object, then
```

dumped into a ``std::ostream``.

Supported Types
^^^^^^^^^^^^^^^

This section provides the serialization of |DynamicData-api| to JSON ostream for all supported types.

.. _xtypes_serialization_utilities_primitive:

Primitives
""""""""""

Primitive types are the basic building blocks for every |DynamicType-api|.
Below is an example of the definition of primitive types in IDL:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_PRIMITIVES
    :end-before: //!--

> The previous |DynamicData-api| object would be serialized as follows:

Reviewer 2: ```suggestion
A |DynamicData-api| object corresponding to the (type representation from above|type represented above|type from above|I think you get it...) would be serialized as follows:
```

Reviewer 2: Please change elsewhere.

.. _xtypes_serialization_utilities:

Serialization Utilities
=======================

Fast DDS provides methods to serialize `XTypes` objects, enabling efficient data exchange and manipulation
within distributed systems.
Serialization is a crucial process in data distribution services, as it converts complex data structures into a
format that can be easily transmitted and reconstructed across different platforms and programming environments.

DynamicData to JSON
--------------------

In the context of DDS (Data Distribution Service), |DynamicType-api| represents the structure of the data being
distributed across the system.
Each |DynamicData-api| object corresponds to an object of the type represented by its |DynamicType-api|,
providing functionalities to access and modify data values.
To enhance interoperability and readability, it is often useful to serialize |DynamicData-api| into a more
manageable format, to enable easier data processing and analysis across different systems and applications.
The method |XTypesUtils-json_serialize-api| converts a |DynamicData-api| object into a JSON object, then
dumped into a ``std::ostream``.

> Supported Types

Reviewer 2: Missing bitsets.

.. include:: ../../03-exports/aliases-api.include

.. _xtypes_type_serializing_json:

> Dynamic Type Serializing

Reviewer 2: How about Serialization helpers/utils/utilities

========================

Fast-DDS provides methods to serialize `DynamicType` objects.

Dynamic Data to JSON
--------------------

> The method `json_serialize` converts a `DynamicData` object into an JSON string.

Reviewer 2: Update as the string overload has been removed.

The following `DynamicData` object would be serialized as follows:

.. literalinclude:: /../code/DynamicTypesIDLExamples.idl
    :language: omg-idl
    :start-after: //!--IDL_TYPE_INTROSPECTION_EXAMPLE
    :end-before: //!--

.. literalinclude:: /../code/DynamicDataJSONExamples.json
    :language: json

Example: Remote discovered types introspection
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following code demonstrates how to implement remote type introspection using FastDDS in C++.

> This feature allows a subscriber to introspect the remotly discovered type and serialize it into a more manageable
> and understandable format.

Reviewer 2: ```suggestion
This feature allows a subscriber to introspect the remotely discovered type and serialize it into a digestible format.
```
